org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/LanguageServerApplication.java

```diff
@@ -42,12 +45,17 @@ public class LanguageServerApplication implements IApplication {
 	private InputStream in;
 	private PrintStream out;
 	private PrintStream err;
+	private ProgressReporterManager progressReporterManager;
 
 	@Override
 	public Object start(IApplicationContext context) throws Exception {
 		prepareWorkspace();
 		prepareStreams();
 		JavaLanguageServerPlugin.startLanguageServer(this);
+		if (JavaLanguageServerPlugin.getInstance().getProtocol() instanceof JDTLanguageServer server) {
```

Comment:
Could this be placed at https://github.com/eclipse/eclipse.jdt.ls/blob/5a7e89f98063624443e85f807723f25969983292/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JavaLanguageServerPlugin.java#L354

```diff
--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JavaLanguageServerPlugin.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JavaLanguageServerPlugin.java
@@ -36,6 +36,7 @@ import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.Platform;
 import org.eclipse.core.runtime.Plugin;
 import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
 import org.eclipse.core.runtime.preferences.DefaultScope;
 import org.eclipse.jdt.core.WorkingCopyOwner;
 import org.eclipse.jdt.core.manipulation.JavaManipulation;
@@ -51,6 +52,7 @@ import org.eclipse.jdt.ls.core.internal.corext.template.java.JavaContextTypeRegi
 import org.eclipse.jdt.ls.core.internal.corext.template.java.JavaLanguageServerTemplateStore;
 import org.eclipse.jdt.ls.core.internal.handlers.CompletionContributionService;
 import org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer;
+import org.eclipse.jdt.ls.core.internal.handlers.ProgressReporterManager;
 import org.eclipse.jdt.ls.core.internal.managers.ContentProviderManager;
 import org.eclipse.jdt.ls.core.internal.managers.DigestStore;
 import org.eclipse.jdt.ls.core.internal.managers.ISourceDownloader;
@@ -351,7 +353,12 @@ public class JavaLanguageServerPlugin extends Plugin {
 			}
 			launcher = Launcher.createLauncher(protocol, JavaLanguageClient.class, in, out, executorService, wrapper);
 		}
-		protocol.connectClient(launcher.getRemoteProxy());
+		JavaLanguageClient client = launcher.getRemoteProxy();
+		protocol.connectClient(client);
+		if (protocol instanceof JDTLanguageServer server) {
+			ProgressReporterManager progressReporterManager = new ProgressReporterManager(client, preferenceManager);
+			Job.getJobManager().setProgressProvider(progressReporterManager);
+		}
 		launcher.startListening();
 	}
 
```

Reply:
Yes, we really shouldn't call it in the API-level, it's Eclipse-application-level setup.

Reply:
Ok, then I would just change it to `server.getClientConnection().client` and just add a comment/todo explaining we want to decouple the initialization of Eclipse specific API from the language server in the future.